### PR TITLE
16 error handling

### PR DESCRIPTION
**Pull request by philippfromme**
Handle errors by rejecting.

### CODE REVIEW

File: test/spec/migrateDiagramSpec.js [hunk 1]
```diff
@@ -183,12 +143,169 @@ describe('migrateDiagram', function() {
...
+      // given
+      const xml = read('test/fixtures/text.txt');
+
+      const migrate = () => {
+        return migrateDiagram(xml);
+      };
+
+      // when
+      // then
+      migrate().should.be.rejectedWith('expected DMN diagram, found FOO');
```
[the remark this answers is not in the document]
Comment (barmac) [reply test/spec/migrateDiagramSpec.js]: Come on, you surely know it's gonna print a two megabyte text document :D

File: src/migrateDiagram.js [hunk 2]
```diff
@@ -29,16 +29,25 @@ const DMN13URI = '"https://www.omg.org/spec/DMN/20191111/MODEL/"';
...
-  } else if (hasNamespace(DMN11URI, xml)) {
-    return migrateFrom11To13(xml);
-  } else if (hasNamespace(DMN13URI, xml)) {
-    return Promise.resolve(xml);
-  }
+    if (hasNamespace(DMN11URI, xml)) {
+      try {
+        resolve(migrateFrom11To13(xml));
+      } catch (err) {
+        reject(err);
```
Comment (barmac) [inline src/migrateDiagram.js]:
```suggestion
        return reject(err);
```

File: src/migrateDiagram.js [hunk 3]
```diff
@@ -29,16 +29,25 @@ const DMN13URI = '"https://www.omg.org/spec/DMN/20191111/MODEL/"';
...
+      }
+    } else if (hasNamespace(DMN12URI, xml)) {
+      resolve(migrateFrom12To13(xml));
+    } else if (hasNamespace(DMN13URI, xml)) {
+      resolve(xml);
+    }
 
-  throw new Error('unknown namespace');
+    reject(new Error(`expected DMN diagram, found ${ xml }`));
```
Comment (barmac) [inline src/migrateDiagram.js]: We always end up here due to missing return statements.

**Comment by philippfromme** [comment]
> * do only return diagrams if you actually migrated

So if I'd pass a DMN 1.3 diagram I'd receive null?

File: src/migrateDiagram.js (hunk 3 above)
Comment (philippfromme) [reply src/migrateDiagram.js]: 💩

File: src/migrateDiagram.js [hunk 4]
```diff
@@ -29,16 +29,25 @@ const DMN13URI = '"https://www.omg.org/spec/DMN/20191111/MODEL/"';
...
+  });
```
Comment (barmac) [inline src/migrateDiagram.js]: BTW this whole function could be async

```suggestion
  if (!isString(xml)) {
    throw new Error(`expected DMN diagram, found ${ xml }`);
  }

  if (hasNamespace(DMN11URI, xml)) {
    return await migrateFrom11To13(xml);
  } else if (hasNamespace(DMN12URI, xml)) {
    return await migrateFrom12To13(xml);
  } else if (hasNamespace(DMN13URI, xml)) {
    return xml;
  }

  throw new Error(`expected DMN diagram, found ${ xml }`);
```

**Comment by philippfromme** [comment]
My suggestion would be: Always return either input or migrated diagram as long as there was no error during the actual migration part.

File: src/migrateDiagram.js (hunk 4 above)
Comment (barmac) [reply src/migrateDiagram.js]: This suggestion thing seems to be broken :D

Comment (philippfromme) [reply src/migrateDiagram.js]: As discussed with @nikku we're not going to use `async`. I think the reasoning was that the Camunda products using this module need to support older browsers. Alternatively we could build distros for these browsers.

**Comment by barmac** [comment]
> > * do only return diagrams if you actually migrated
> 
> So if I'd pass a DMN 1.3 diagram I'd receive null?

That could be a problem for us when you batch migrate a whole project.

File: src/migrateDiagram.js (hunk 4 above)
Comment (barmac) [reply src/migrateDiagram.js]: All right, I get it.

**Comment by nikku** [comment]
> My suggestion would be: Always return either input or migrated diagram as long as there was no error during the actual migration part.

Fine with that.

**Comment by barmac** [comment]
> > My suggestion would be: Always return either input or migrated diagram as long as there was no error during the actual migration part.
> 
> Fine with that.

🇫🇮 same here

**Comment by philippfromme** [comment]
Cool. I think that would result in less friction. Things can still blow up on import. Before that, only blow up if the migration itself failed.

**Comment by nikku** [comment]
:cool: :iceland:

**Comment by philippfromme** [comment]
Ready for review - again. ♻️

File: test/spec/migrateDiagramSpec.js [hunk 5]
```diff
@@ -183,12 +143,117 @@ describe('migrateDiagram', function() {
...
+      migrate().should.be.rejectedWith('Cannot set property \'$parent\' of undefined');
+    });
+
+  });
+
 });
 
 
 
-// helper ///////////////
```
Comment (nikku) [inline test/spec/migrateDiagramSpec.js]: :o:

**Pull request merged**